src/NuGet.Core/NuGet.Packaging/TopologicalSortUtility.cs

```diff
@@ -10,24 +10,21 @@ namespace NuGet.Packaging
 {
     public static class TopologicalSortUtility
     {
+        private static readonly PackageInfoComparer DefaultComparer = new PackageInfoComparer();
         /// <summary>
         /// Order dependencies by children first.
         /// </summary>
         public static IReadOnlyList<PackageDependencyInfo> SortPackagesByDependencyOrder(
             IEnumerable<PackageDependencyInfo> packages)
         {
-            var sorted = new List<PackageDependencyInfo>();
-            var toSort = packages.Distinct().ToList();
+            var toSort = packages.Distinct().Select(package => new PackageInfo(package)).ToArray();
```

Comment:
You mean the `Select` transform? That's to add `ParentCount` as a property for caching the parent count so it only needs to be allocated once per function call, rather than a per loop key and map allocation via Linq.Sort (which is still better than a pure functional compare with parent recalc per sort element, without caching the values)

Also since `PackageInfo` is a struct; as a bonus, it changes stops the underlying array (in List or direct) from being covariant and having per op checks, for an extra boost 😉